### PR TITLE
[BREAKING] C99 scoping rules by default (remove 050 workaround)

### PR DESCRIPTION
**Pull request by leonardoalt**

### CODE REVIEW

**Comment by erak** [comment]
This needs a Changelog entry as well. And is there any documentation we need to update here?

File: test/libsolidity/syntaxTests/double_variable_declaration.sol [hunk 1]
```diff
@@ -5,4 +5,6 @@ contract test {
 	}
 }
 // ----
-// DeclarationError: (71-80): Identifier already declared.
+// Warning: (71-80): This declaration shadows an existing declaration.
+// Warning: (46-55): Unused local variable.
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/double_variable_declaration.sol]: You may want to get rid of these warnings by using ``x;`` after both declarations. Fine either way, though.

File: test/libsolidity/syntaxTests/nameAndTypeResolution/448_no_unused_dec_after_use.sol [hunk 2]
```diff
@@ -4,3 +4,5 @@ contract C {
         uint a;
     }
 }
+// ----
+// DeclarationError: (52-53): Undeclared identifier. Did you mean "a"?
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/448_no_unused_dec_after_use.sol]: ``Did you mean "a"?`` is rather misleading here - maybe we should change or remove that suggestion.

File: test/libsolidity/syntaxTests/parsing/conditional_with_assignment.sol [hunk 3]
```diff
@@ -1,11 +1,10 @@
 contract A {
     function f() {
         uint y = 1;
-        uint x = 3 < 0 ? x = 3 : 6;
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/parsing/conditional_with_assignment.sol]: Maybe there should be another test about this not working with ``x``?

File: test/libsolidity/syntaxTests/parsing/conditional_with_assignment.sol [hunk 4]
```diff
@@ -1,11 +1,10 @@
...
+        uint x = 3 < 0 ? y = 3 : 6;
         true ? x = 3 : 4;
     }
 }
 // ----
 // Warning: (17-119): No visibility specified. Defaulting to "public".
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/parsing/conditional_with_assignment.sol]: We could also get rid of the (unrelated) visibility and mutability warnings while we're at it. Same for some other tests. But fine in any case.

File: test/libsolidity/syntaxTests/parsing/while_loop.sol [hunk 5]
```diff
@@ -1,9 +1,10 @@
...
     function fun(uint256 a) {
-        while (true) { uint256 x = 1; break; continue; } x = 9;
+        uint256 x;
+        while (true) { x = 1; break; continue; } x = 9;
     }
 }
 // ----
-// Warning: (20-115): No visibility specified. Defaulting to "public". 
+// Warning: (20-126): No visibility specified. Defaulting to "public". 
 // Warning: (33-42): Unused function parameter. Remove or comment out the variable name to silence this warning.
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/parsing/while_loop.sol]: Why does ``fun`` have an argument for this test anyways? Doesn't seem related to or necessary for the test...

File: test/libsolidity/syntaxTests/scoping/scoping_self_use.sol [hunk 6]
```diff
@@ -3,3 +3,5 @@ contract test {
         uint a = a;
     }
 }
+// ----
+// DeclarationError: (64-65): Undeclared identifier. Did you mean "a"?
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/scoping/scoping_self_use.sol]: Here the suggestion is even stranger than in the other cases.

File: test/libsolidity/syntaxTests/parsing/conditional_with_assignment.sol (hunk 3 above)
Comment (ekpyron) [reply test/libsolidity/syntaxTests/parsing/conditional_with_assignment.sol]: Actually, not sure why there was a ``y`` in the original test - maybe this was always meant to be ``y`` anyways...

**Review by ekpyron — COMMENTED** [review]
For now mostly some suggestions that actually may not need to be in this PR.

File: test/libsolidity/syntaxTests/parsing/conditional_with_assignment.sol (hunk 4 above)
Comment (axic) [reply test/libsolidity/syntaxTests/parsing/conditional_with_assignment.sol]: I'd do that in a separate PR, since it should also make that an error and not a warning - though I am not sure we made a final decision on that yet, but we were supposed to make that mandatory.

File: Changelog.md [hunk 7]
```diff
@@ -19,6 +19,7 @@ Breaking Changes:
  * Type Checker: Disallow arithmetic operations for boolean variables.
  * Type Checker: Disallow conversions between ``bytesX`` and ``uintY`` of different size.
  * Remove obsolete ``std`` directory from the Solidity repository. This means accessing ``https://github.com/ethereum/soldity/blob/develop/std/*.sol`` (or ``https://github.com/ethereum/solidity/std/*.sol`` in Remix) will not be possible.
+ * General: C99 scoping rules.
```
Comment (axic) [inline Changelog.md]: Should be extended along the lines of:  "C99-style are enforced now. This was already the case in the experimental 0.5.0 mode."

File: docs/control-structures.rst [hunk 8]
```diff
@@ -325,74 +325,7 @@ is ``false``. The default value for the ``uint`` or ``int`` types is ``0``. For
...
-            }
-            return bar;// returns 5
-        }
-    }
-
-Scoping starting from Version 0.5.0
------------------------------------
-
-Starting from version 0.5.0, Solidity will change to the more widespread scoping rules of C99
+Starting from version 0.5.0, scoping in Solidity follows the more widespread scoping rules of C99
```
Comment (axic) [inline docs/control-structures.rst]: I think instead of "Starting from version" we should just make this read as "Solidity follows the widespread scoping rules of C99..." and add a "warning" section below: "Before version 0.5.0 Solidity has follows scoping rules of JavaScript." (maybe more explanation can be added if required).

File: docs/control-structures.rst [hunk 9]
```diff
@@ -401,17 +334,12 @@ Variables and other items declared outside of a code block, for example function
...
-
 As a consequence, the following examples will compile without warnings, since
-the two variables have the same name but disjoint scopes. In non-0.5.0-mode,
-they have the same scope (the function ``minimalScoping``) and thus it does
-not compile there.
+the two variables have the same name but disjoint scopes.
 
 ::
 
     pragma solidity ^0.4.0;
```
Comment (axic) [inline docs/control-structures.rst]: Need to update this.

File: docs/control-structures.rst [hunk 10]
```diff
@@ -431,7 +359,6 @@ In any case, you will get a warning about the outer variable being shadowed.
...
```
Comment (axic) [inline docs/control-structures.rst]: Same here.

File: libsolidity/analysis/ReferencesResolver.cpp [hunk 11]
```diff
@@ -101,7 +93,11 @@ bool ReferencesResolver::visit(Identifier const& _identifier)
 		string suggestions = m_resolver.similarNameSuggestions(_identifier.name());
 		string errorMessage =
 			"Undeclared identifier." +
-			(suggestions.empty()? "": " Did you mean " + std::move(suggestions) + "?");
+			(suggestions.empty()? "":
+				(string("\"" + _identifier.name() + "\"") != suggestions)?
+					(" Did you mean " + std::move(suggestions) + "?"):
```
Comment (axic) [inline libsolidity/analysis/ReferencesResolver.cpp]: Style: space before `:`

File: test/compilationTests/corion/provider.sol [hunk 12]
```diff
@@ -585,8 +586,9 @@ contract provider is module, safeMath, announcementTypes {
         uint256 steps;
         uint256 currHeight = providers[addr].currentHeight;
         uint256 LTSID = providers[addr].data[currHeight].lastSupplyID;
+		uint256 a;
```
Comment (axic) [inline test/compilationTests/corion/provider.sol]: Style: indentation

File: test/compilationTests/stringutils/strings.sol [hunk 13]
```diff
@@ -688,7 +689,8 @@ library strings {
             return "";
 
         uint len = self._len * (parts.length - 1);
-        for(uint i = 0; i < parts.length; i++)
+        uint i;
```
Comment (axic) [inline test/compilationTests/stringutils/strings.sol]: While this is not crucial, probably better style is to keep the variable declaration in each for loop.

File: test/libsolidity/syntaxTests/modifiers/function_modifier_invocation_local_variables.sol [hunk 14]
```diff
@@ -2,3 +2,5 @@ contract B {
     function f() mod(x) pure public { uint x = 7; }
     modifier mod(uint a) { if (a > 0) _; }
 }
+// ----
+// DeclarationError: (34-35): Undeclared identifier.
```
Comment (axic) [inline test/libsolidity/syntaxTests/modifiers/function_modifier_invocation_local_variables.sol:6]: Hm. This perhaps should be noted in the "breaking changes" section. We do not have one in the documentation yet.., but should be in the changelog.

File: test/libsolidity/syntaxTests/nameAndTypeResolution/448_no_unused_dec_after_use.sol [hunk 15]
```diff
@@ -4,3 +4,5 @@ contract C {
         uint a;
```
Comment (axic) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/448_no_unused_dec_after_use.sol]: Duplicate of scoping/scoping_activation_old.sol. The other one looks more relevantly grouped,  I'd delete this one.

File: test/libsolidity/syntaxTests/modifiers/function_modifier_invocation_local_variables.sol (hunk 14 above)
Comment (leonardoalt) [reply test/libsolidity/syntaxTests/modifiers/function_modifier_invocation_local_variables.sol:6]: Should we have it in both `Language features` and `breaking changes` sections in the Changelog?

**Comment by leonardoalt** [comment]
Fixed review suggestions, squashed and rebased.

File: test/libsolidity/syntaxTests/modifiers/function_modifier_invocation_local_variables.sol (hunk 14 above)
Comment (axic) [reply test/libsolidity/syntaxTests/modifiers/function_modifier_invocation_local_variables.sol:6]: Yes, both. Plus we should really have a section in the documentation I think - we can do that in a separate PR though.

Comment (leonardoalt) [reply test/libsolidity/syntaxTests/modifiers/function_modifier_invocation_local_variables.sol:6]: Done

File: docs/control-structures.rst [hunk 16]
```diff
@@ -443,6 +370,9 @@ In any case, you will get a warning about the outer variable being shadowed.
         }
     }
 
+.. warning::
+    Before version 0.5.0 Solidity followed the same scoping rules as JavaScript.
```
Comment (chriseth) [inline docs/control-structures.rst]: Could you extend this section a little? Most people think that JavaScript has C99 scoping rules.

File: libsolidity/analysis/ReferencesResolver.cpp [hunk 17]
```diff
@@ -101,7 +93,11 @@ bool ReferencesResolver::visit(Identifier const& _identifier)
 		string suggestions = m_resolver.similarNameSuggestions(_identifier.name());
 		string errorMessage =
 			"Undeclared identifier." +
-			(suggestions.empty()? "": " Did you mean " + std::move(suggestions) + "?");
```
Comment (chriseth) [inline libsolidity/analysis/ReferencesResolver.cpp]: This gets rather complicated with the ternary operators. Perhaps it is better to construct `errorMessage` in multiple statements and use `if` instead?

Comment (chriseth) [reply libsolidity/analysis/ReferencesResolver.cpp]: Also spacing: Exactly one space before and after `?` and `:` and usually we do not use additional parentheses.

File: test/libsolidity/syntaxTests/nameAndTypeResolution/137_external_visibility.sol [hunk 18]
```diff
@@ -3,4 +3,4 @@ contract c {
     function g() public { f(); }
 }
 // ----
-// DeclarationError: (68-69): Undeclared identifier. Did you mean "f"?
+// DeclarationError: (68-69): Undeclared identifier. Did you mean to declare "f" earlier?
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/137_external_visibility.sol]: In this case, the error is misleading. Perhaps we should rather change the message to `"f" is not (or not yet) visible at this point.`

File: test/libsolidity/syntaxTests/nameAndTypeResolution/448_no_unused_dec_after_use.sol [hunk 19]
```diff
@@ -1,6 +0,0 @@
-contract C {
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/448_no_unused_dec_after_use.sol:1]: Do we still have a test like this?

Comment (chriseth) [reply test/libsolidity/syntaxTests/nameAndTypeResolution/448_no_unused_dec_after_use.sol:1]: Ah ok, we have! :)

**Review by chriseth — CHANGES_REQUESTED** [review]
Please change the error message (both its wording and how it is constructed) and extend the documentation about the old scoping rules.

**Review by chriseth — DISMISSED** [review]

**Review by chriseth — APPROVED** [review]
